### ISSUE TITLE
Simplified sampler for reduced USB bandwidth usage

### ISSUE THREAD

**Issue by Bonstra** [davidgiven/fluxengine#141]
I've written a proof of concept implementation of a simplified Sampler, which uses a 6 bit counter instead of 7, allowing to pack every event in a single byte, thus removing the need for crunching data.

I can now read HD 3.5" IBM-formatted floppies without issue, when I couldn't before (attempting to read tracks 0, 1 and 2 would fail (amusingly, the other tracks read fine)).

The code is available here for testing: https://github.com/Bonstra/fluxengine/commits/simplified-sampler
All the details of what changed are in the commit messages.

I simulated the Sampler verilog in Verilator to ensure it behaves correctly, and it does*.

*as long as you keep the sample clock exactly 3 times slower than the bus clock :-p

I'm not making a pull request for this branch yet because:
- it is based on a rather old commit, and you can't just rebase commits making changes in a PSoC Creator project file…
- it contains quick and ugly hacks to make it work without refactoring too much
- it breaks features: write support, INDEX300 and INDEX360 pulses
- it contains some unrelated changes which were not removed from the commits because of the binary format used by PSoC Creator
- a proper implementation would require changing the format used internally by the fluxengine client to process the flux data (hence the ugly hacks to work around that and the broken write support)

If you believe it's worth doing, I can manually re-apply the cleaned-up changes on the latest release, and fix the sample clock bug in the verilog.

**Comment by davidgiven** [comment]
Thank you very much for that!

However... you'll never guess what I just spent several weeks doing...

Okay, so I didn't switch to a 6-bit timer, but I did rewrite the sampler and sequencer (several times) while also rewriting the cruncher library for USB compression. Bandwidth is still pretty high and you need 800-900 kB/s to read or write an HD disk. But then we always did; there just happened to be a bug in the firmware where it didn't block properly before reusing USB buffers, which meant that at high bandwidths you got mangled packets. It's a miracle it worked at all.

I did think of a six-bit timer but never actually did the maths, assuming that wasting two bits per byte would push the bandwidth too high --- instead I relied on the cruncher to encode the data in software. The new cruncher will in fact transfer a pulse and an interval in 8 bits where possible, and 11 when not: https://github.com/davidgiven/fluxengine/blob/master/lib/common/crunch.h

For an HD disk every interval is shorter than 64 ticks, so each one would be encoded as a single byte --- obviously. However a 720kB PC disk tends to have intervals all the way up to 8us, or 100 ticks These would be encoded as two bytes in the new scheme, so that's actually more than the current crunched bytestream. I think it would still end up being a win here because most intervals are 4us, or 50 ticks --- these would be encoded as a single byte. A Brother disk, however, has the majority of intervals at 7.42us, or 92 ticks. Each one of these will be two bytes.

My maths (and bear in mind it's 23:00 here) suggests that all the bandwidths come out at about 300-500kB, which would be great. _However_ that should be true for the current cruncher scheme too, and it's not, so that probably wants investigating.

I'm inclined to think that your proposal should win for sheer simplicity alone. It would allow big chunks to be removed from the firmware, including the cruncher, and getting that USB stuff is hard enough without two different producer/consumer chains working at once. There would be no need to change the PC data representation, as the bytestream could be transparently converted either way, but that might actually have merit too due to being easier to handle. (No more opcodes --- you simply ask the FluxReader for the next event and get a bitmask of index|pulse|end of stream.)

**Comment by davidgiven** [comment]
The PR at #146 reworks the bytecode format to the proposed (or at least _a_ proposed, I was on a plane and didn't have access to your branch) new format. I haven't touched the firmware yet, though.

**Comment by Bonstra** [comment]
In my tests, the bandwidth issues I had with reading HD disk were caused by the software cruncher not crunching fast enough to fill the USB EP buffer.
At first, I had started making a verilog implementation of the cruncher, which was made irrelevant by the counter width reduction :P
I've given your PR a try, but couldn't get it to work. I think I messed up something in the firmware. I'll give it another shot within the week.

**Comment by davidgiven** [comment]
Yes, in hindsight the cruncher overrunning explains all my symptoms...

After way, way too much work (and being on holiday) I've pulled out the crunch code and updated the firmware so it'll read (but not write) the new format. Everything seems to work in practice.

The way too much work was because I spent several days trying to figure out what I was doing wrong which was causing the sampled data to be massively corrupted --- it looked like two concurrent data streams in alternate bytes. It turns out it's _always_ been massively corrupted, it's just that with the old format every other byte was always a 0x80 so it didn't matter if they got interchanged! Changing the DMA settings to use one-byte bursts rather than two-byte bursts fixed that. I think it was triggering a two-byte read as soon as the FIFO contained a single byte, but I'm still not sure why it wasn't causing FIFO underruns. So, thanks again for making me look here in more detail. That probably explains several unexplained issues.

Next step is the sequencer.

**Comment by Bonstra** [comment]
I can confirm that reading now works, even with the problematic disk which was causing USB bandwidth issues in HD mode.
The DMA burst size issue is probably the reason why I couldn't get your PR to work last time. That is not something I suspected.
Thank you for your work!

**Comment by davidgiven** [comment]
It's taken a while but I believe I now have the sequencer working too (although it currently ignores index bits --- I'll handle then when I actually have a hard-sectored format which requires them). There were a bunch of USB issues which needed fixing which have, I hope, been fixed, but I don't quite understand why there were problems in the first place; sometimes the first packet of a transfer would be dropped, but only on writes. The PSOC USB library isn't the friendliest. There's still something which needs fixing which causes occasional timeouts on command transfers on Linux.

The other is written disks don't work very well. My suspicion is that the PSoC's sampleclock isn't quite within tolerance (IBM HD disks are reported at 512kHz on my device, rather than 500kHz).

**Comment by davidgiven** [comment]
I have merged the fix for this; it's working really well. Thank you very much. Please let me know if you want a credit, and if so, what as.

Unfortunately there seems to still be something wrong with the write support, but I don't think it's an issue with the sequencer or the firmware: it seems to manifest as the clock running slow, but to the best of my testing the hardware's fine. I have some parts on order which will let me test this once and for all. I think it was always broken, and was just working by accident, hence the merge.
